src/charm.py

```diff
@@ -549,11 +556,18 @@ def _on_peer_relation_changed(self, event: HookEvent) -> None:  # noqa: C901
             logger.error("Invalid configuration: %s", str(e))
             return
 
-        # If PITR restore failed, then wait it for resolve.
+        # Should not override a blocked status
+        if isinstance(self.unit.status, BlockedStatus):
+            logger.debug("on_peer_relation_changed early exit: Unit in blocked status")
+            return
+
+        services = container.pebble.get_services(names=[self._postgresql_service])
         if (
-            "restoring-backup" in self.app_peer_data or "restore-to-time" in self.app_peer_data
-        ) and isinstance(self.unit.status, BlockedStatus):
-            event.defer()
+            ("restoring-backup" in self.app_peer_data or "restore-to-time" in self.app_peer_data)
+            and len(services) > 0
+            and not self._was_restore_successful(container, services[0])
+        ):
+            logger.debug("on_peer_relation_changed early exit: Backup restore check failed")
```

Comment:
It was observed in some cases that, after a PITR restore failed, the first event to run following the failure would be `on_peer_relation_changed` instead of `update_status` (where backup failures are normally spotted by the charm). When that happened, the charm would try to ping patroni, fail, enter `awaiting for member to start` status and endlessly defer, never knowing it was a backup failure. this condition is meant to avoid that.